File: clients/python/beneath/checkpointer.py
# allows us to use Client as a type hint without an import cycle
# see: https://www.stefaanlippens.net/circular-imports-type-hints-python.html
# pylint: disable=wrong-import-position,ungrouped-imports
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from beneath.client import Client

from datetime import timedelta
import json
import sys
from typing import Any, Dict, Tuple

import msgpack

from beneath.config import DEFAULT_CHECKPOINT_COMMIT_DELAY_MS
from beneath.instance import StreamInstance
from beneath.utils import AIODelayBuffer, StreamQualifier

SERVICE_CHECKPOINT_LOG_RETENTION = timedelta(hours=6)
SERVICE_CHECKPOINT_SCHEMA = """
  type Checkpoint @schema {
    key: String! @key
    value: Bytes
  }
"""


class Checkpointer:
    """
    Checkpointers store (small) key-value records in a meta stream (in the specified project).
    They are useful for maintaining consumer and pipeline state, such as the cursor for a
    subscription or the last time a scraper ran.

    Checkpoint keys are strings and values are serialized with msgpack (supports most normal Python
    values, but not custom classes).

    New checkpointed values are flushed at regular intervals (every 30 seconds by default).
    Checkpointers should not be used for storing large amounts of data.
    Checkpointers are not currently suitable for synchronizing parallel processes.
    """

    instance: StreamInstance
    """ The meta-stream instance that checkpoints are written to """

    def __init__(
        self,
        client: Client,
        metastream_qualifier: StreamQualifier,
        metastream_create: bool,
        metastream_description: str,
    ):
        self.instance = None
        self._client = client
        self._metastream_qualifier = metastream_qualifier
        self._metastream_description = metastream_description
        self._create = metastream_create
        self._writer = self._Writer(self)
        self._cache: Dict[str, Any] = {}

    async def get(self, key: str, default: Any = None) -> Any:
        """
        Gets a checkpointed value
        """
        if key in self._cache:
            return self._cache[key]

        if self._client.dry:
            return default

        filt = json.dumps({"key": key})
        cursor = await self.instance.query_index(filter=filt)

        value = default
        record = await cursor.read_one()
        if record is not None:
            value = msgpack.unpackb(record["value"], timestamp=3)

        # checking self._cache again because of awaits (and we'd rather serve a recent local set)
        if key not in self._cache:
            self._cache[key] = value

        return self._cache[key]

    async def set(self, key: str, value: Any):
        """ Sets a checkpoint value. Value will be encoded with msgpack. """
        if not self._writer.running:
            raise Exception("Cannot call 'set' on checkpointer because the client is stopped")
        self._cache[key] = value
        await self._writer.write(key, value)

    # START/STOP (called by client)

    async def _start(self):
        if not self.instance:
            await self._stage_stream()
        await self._writer.start()

    async def _stop(self):
        await self._writer.stop()

    # CHECKPOINT STREAM

    async def _stage_stream(self):
        if self._create or self._client.dry:
            stream = await self._client.create_stream(
                stream_path=str(self._metastream_qualifier),
                schema=SERVICE_CHECKPOINT_SCHEMA,
                description=self._metastream_description,
                meta=True,
                log_retention=SERVICE_CHECKPOINT_LOG_RETENTION,
                use_warehouse=False,
                update_if_exists=True,
            )
        else:
            stream = await self._client.find_stream(stream_path=str(self._metastream_qualifier))

        if not stream.primary_instance:
            raise Exception("Expected checkpoints stream to have a primary instance")
        self.instance = stream.primary_instance

        self._client.logger.info(
            "Using '%s' (version %i) for checkpointing",
            self._metastream_qualifier,
            self.instance.version,
        )

    # CHECKPOINT WRITER

    class _Writer(AIODelayBuffer[Tuple[str, Any]]):

        checkpoint: Dict[str, Any]

        def __init__(self, checkpointer: Checkpointer):
            super().__init__(
                max_delay_ms=DEFAULT_CHECKPOINT_COMMIT_DELAY_MS,
                max_record_size=sys.maxsize,
                max_buffer_size=sys.maxsize,
                max_buffer_count=sys.maxsize,
            )
            self.checkpointer = checkpointer

        def _reset(self):
            self.checkpoint = {}

        def _merge(self, value: Tuple[str, Any]):
            (key, checkpoint) = value
            self.checkpoint[key] = checkpoint

        async def _flush(self):
            records = (
                {
                    "key": key,
                    "value": msgpack.packb(checkpoint, datetime=True),
                }
                for (key, checkpoint) in self.checkpoint.items()
            )
            if self.checkpointer.instance:
                await self.checkpointer._client.write(
                    instance=self.checkpointer.instance,
                    records=records,
                )

        # pylint: disable=arguments-differ
        async def write(self, key: str, checkpoint: Any):
            await super().write(value=(key, checkpoint), size=0)


class PrefixedCheckpointer:
    """ Wraps a Checkpointer and prefixes all keys """

    def __init__(self, checkpointer: Checkpointer, prefix: str):
        self._checkpointer = checkpointer
        self._prefix = prefix

    @property
    def instance(self):
        return self._checkpointer.instance

    async def get(self, key: str, default: Any = None) -> Any:
        key = self._prefix + key
        return await self._checkpointer.get(key, default=default)

    async def set(self, key: str, value: Any):
        key = self._prefix + key
        return await self._checkpointer.set(key, value)

File: clients/java/src/main/java/dev/beneath/client/Checkpointer.java
package dev.beneath.client;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import org.apache.avro.Schema;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.generic.GenericRecordBuilder;
import org.msgpack.jackson.dataformat.MessagePackFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import dev.beneath.client.utils.AIODelayBuffer;
import dev.beneath.client.utils.TableIdentifier;
import dev.beneath.type.TableSchemaKind;

/**
 * Checkpointers store (small) key-value records in a meta table (in the
 * specified project). They are useful for maintaining consumer and pipeline
 * state, such as the cursor for a subscription or the last time a scraper ran.
 * 
 * Checkpoint keys are strings and values are serialized with msgpack (supports
 * most normal Python values, but not custom classes).
 * 
 * New checkpointed values are flushed at regular intervals (every 30 seconds by
 * default). Checkpointers should not be used for storing large amounts of data.
 * Checkpointers are not currently suitable for synchronizing parallel
 * processes.
 */
public class Checkpointer {
  public TableInstance instance;
  private BeneathClient client;
  private TableIdentifier metatableIdentifier;
  private String metatableDescription;
  private Boolean create;
  private Writer writer;
  private Map<String, Object> cache;
  private ObjectMapper objectMapper;
  private static final Integer SERVICE_CHECKPOINT_LOG_RETENTION = 60 * 60 * 6; // 6 hours
  private static final String SERVICE_CHECKPOINT_SCHEMA = """
      type Checkpoint @schema {
      key: String! @key
      value: Bytes
      }
      """;
  private static final Logger LOGGER = LoggerFactory.getLogger(Checkpointer.class);

  public Checkpointer(BeneathClient client, TableIdentifier metatableIdentifier, Boolean metatableCreate,
      String metatableDescription) {
    this.client = client;
    this.metatableIdentifier = metatableIdentifier;
    this.metatableDescription = metatableDescription;
    this.create = metatableCreate;
    this.writer = new Writer(this);
    this.cache = new HashMap<String, Object>();
    this.objectMapper = new ObjectMapper(new MessagePackFactory());
  }

  /**
   * Gets a checkpointed value
   */
  public Object get(String key) {
    return this.get(key, null);
  }

  /**
   * Gets a checkpointed value
   */
  public Object get(String key, Object defaultValue) {
    if (this.cache.containsKey(key)) {
      return this.cache.get(key);
    }

    if (this.client.dry) {
      return defaultValue;
    }

    String filter = String.format("{\"key\": \"%s\"}", key);
    Cursor cursor = this.instance.queryIndex(filter);

    Object value = defaultValue;
    GenericRecord record = cursor.readOne();
    if (record != null) {
      byte[] bytes = ((ByteBuffer) record.get("value")).array();
      try {
        value = objectMapper.readValue(bytes, Object.class);
      } catch (IOException e) {
        throw new RuntimeException(e);
      }
    }

    // checking self._cache again because of awaits (and we'd rather serve a recent
    // local set)
    if (!this.cache.containsKey(key)) {
      this.cache.put(key, value);
    }

    return this.cache.get(key);
  }

  /**
   * Sets a checkpoint value. Value will be encoded with msgpack.
   */
  public void set(String key, Object value) {
    if (!this.writer.running) {
      throw new RuntimeException("Cannot call 'set' on checkpointer because the client is stopped");
    }
    this.cache.put(key, value);
    this.writer.write(key, value);
  }

  // START/STOP (called by client)

  public void start() {
    if (this.instance == null) {
      this.stageTable();
    }
    this.writer.start();
  }

  public void stop() {
    this.writer.stop();
  }

  // CHECKPOINT TABLE

  private void stageTable() {
    Table table;
    if (this.create || this.client.dry) {
      table = this.client.createTable(this.metatableIdentifier.toString(), SERVICE_CHECKPOINT_SCHEMA,
          this.metatableDescription, true, true, false, 0, SERVICE_CHECKPOINT_LOG_RETENTION, 0, null,
          TableSchemaKind.GRAPHQL, "", true);
    } else {
      table = this.client.findTable(this.metatableIdentifier.toString());
    }
    if (table.primaryInstance == null) {
      throw new RuntimeException("Expected checkpoints table to have a primary instance");
    }
    this.instance = table.primaryInstance;

    LOGGER.info("Using '{}' (version {}) for checkpointing", this.metatableIdentifier.toString(),
        this.instance.version);
  }

  // CHECKPOINT WRITER

  // probably should make these private and add getters/setters
  class CheckpointKeyValue {
    public String key;
    public Object checkpoint;

    CheckpointKeyValue(String key, Object checkpoint) {
      this.key = key;
      this.checkpoint = checkpoint;
    }
  }

  class Writer extends AIODelayBuffer<CheckpointKeyValue> {
    Checkpointer checkpointer;
    Map<String, Object> checkpoints; // this is the buffer

    Writer(Checkpointer checkpointer) {
      super(Config.DEFAULT_CHECKPOINT_COMMIT_DELAY_MS, Integer.MAX_VALUE, Integer.MAX_VALUE, Integer.MAX_VALUE);
      this.checkpointer = checkpointer;
      this.checkpoints = new HashMap<String, Object>();
    }

    @Override
    protected void reset() {
      if (this.checkpoints != null) {
        this.checkpoints.clear();
      }
    }

    @Override
    protected void merge(CheckpointKeyValue value) {
      this.checkpoints.put(value.key, value.checkpoint);
    }

    @Override
    protected void flush() {
      // iterate through map and serialize w/ messagepack
      List<GenericRecord> records = new ArrayList<GenericRecord>();
      Schema schema = new Schema.Parser().parse(instance.table.schema.parsedAvro.toString());
      for (Entry<String, Object> checkpoint : checkpoints.entrySet()) {
        try {
          ByteBuffer bytes;
          bytes = ByteBuffer.wrap(objectMapper.writeValueAsBytes(checkpoint.getValue()));
          GenericRecord genericRecord = new GenericRecordBuilder(schema).set("key", checkpoint.getKey())
              .set("value", bytes).build();
          records.add(genericRecord);
        } catch (JsonProcessingException e) {
          e.printStackTrace();
        }
      }

      // issue network request
      if (this.checkpointer.instance != null) {
        try {
          this.checkpointer.client.write(this.checkpointer.instance, records);
        } catch (Exception e) {
          e.printStackTrace();
        }
      }
    }

    public void write(String key, Object checkpoint) {
      CheckpointKeyValue checkpointKeyValue = new CheckpointKeyValue(key, checkpoint);
      super.write(checkpointKeyValue, 0);
    }
  }
}
